File: leetcode/Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ret = null;
        ListNode cur = null;
        int c = 0;
        while (l1 != null && l2 != null) {
            int s = l1.val + l2.val + c;
            c = s / 10;
            if (cur == null) {
                ret = cur = new ListNode(s % 10);
            } else {
                cur.next = new ListNode(s % 10);
                cur = cur.next;
            }
            l1 = l1.next;
            l2 = l2.next;
        }
        while (l1 != null) {
            int s = l1.val + c;
            c = s / 10;
            if (cur == null) {
                ret = cur = new ListNode(s % 10);
            } else {
                cur.next = new ListNode(s % 10);
                cur = cur.next;
            }
            l1 = l1.next;
        }
        while (l2 != null) {
            int s = l2.val + c;
            c = s / 10;
            if (cur == null) {
                ret = cur = new ListNode(s % 10);
            } else {
                cur.next = new ListNode(s % 10);
                cur = cur.next;
            }
            l2 = l2.next;
        }
        if (c > 0) {
            if (cur == null) {
                ret = cur = new ListNode(c);
            } else {
                cur.next = new ListNode(c);
                cur = cur.next;
            }
        }
        return ret;
    }
}

File: leetcode/Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *head = NULL;
        ListNode *prev = NULL;
        int inc = 0;
        while (l1 != NULL && l2 != NULL) {
            int sum = l1->val + l2->val + inc;
            inc = sum / 10;
            if (head == NULL) {
                prev = head = new ListNode(sum % 10);
            } else {
                prev->next = new ListNode(sum % 10);
                prev = prev->next;
            }
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1 != NULL) {
            int sum = l1->val + inc;
            inc = sum / 10;
            prev->next = new ListNode(sum % 10);
            prev = prev->next;
            l1 = l1->next;
        }
        while (l2 != NULL) {
            int sum = l2->val + inc;
            inc = sum / 10;
            prev->next = new ListNode(sum % 10);
            prev = prev->next;
            l2 = l2->next;
        }
        if (inc > 0)
            prev->next = new ListNode(inc);
        return head;
    }
};
